fvm/state/state_holder.go

```diff
@@ -16,29 +30,226 @@ import (
-func (s *StateHolder) State() *State {
-	return s.activeState
+// CommitToParent commits the changes in the current unrestricted nested
+// transaction to the parent (nested) transaction.
+func (s *StateHolder) CommitToParent() error {
+	if s.IsParseRestricted() {
+		// This is due to a programming error.
+		return fmt.Errorf(
+			"cannot commit unexpected/restricted nested transaction",
+		)
+	}
+
+	return s.mergeIntoParent(s.EnforceInteractionLimits())
 }
 
-// TODO(patrick): remove
-// SetActiveState sets active state
-func (s *StateHolder) SetActiveState(st *State) {
-	s.activeState = st
+// CommitToMain ensures there's only a single outstanding unrestricted nested
+// transaction and commits the changes to the main transaction.
+func (s *StateHolder) CommitToMain() error {
+	if s.NumNestedTransactions() > 1 {
+		return fmt.Errorf("found too many nested transactions")
+	}
+
+	return s.CommitToParent()
+}
+
+// CommitParseRestrictedToParent commits the changes in the current restricted
+// nested transaction to the parent (nested) transaction.  This returns
+// error if the specified location does not match the tracked location.
+func (s *StateHolder) CommitParseRestrictedToParent(
+	location common.AddressLocation,
+) (
+	*State,
+	error,
+) {
+	currentFrame := s.current()
+	if currentFrame.parseRestriction == nil ||
+		currentFrame.parseRestriction.ID() != location.ID() {
+
+		// This is due to a programming error.
+		return nil, fmt.Errorf(
+			"cannot commit unexpected nested transaction %v != %v",
+			currentFrame.parseRestriction,
+			location,
+		)
+	}
+
+	err := s.mergeIntoParent(s.EnforceInteractionLimits())
+	if err != nil {
+		return nil, err
+	}
+
+	return currentFrame.state, nil
+}
+
+// AttachAndCommitNestedTransactionForParsingProgram commits the changes in the
+// cached nested transaction to the current (nested) transaction.
+func (s *StateHolder) AttachAndCommitParseRestrictedToParent(
+	cachedNestedTransaction *State,
+) error {
+	s.nestedTransactions = append(
+		s.nestedTransactions,
+		nestedTransactionStackFrame{
+			state:            cachedNestedTransaction,
+			parseRestriction: nil,
+		},
+	)
+
+	// NOTE: limit enforcement is disabled here because cadence environment's
+	// Get() does not support returning error.
+	return s.mergeIntoParent(false)
+}
+
+// RestartNestedTransaction merges all changes that belongs to the nested
+// transaction about to be restart (for spock/meter bookkeeping), then
+// wipes its view changes.
+func (s *StateHolder) RestartNestedTransaction(
+	id NestedTransactionId,
+) error {
+	found := false
+	for _, frame := range s.nestedTransactions {
+		if frame.state == id.state {
+			found = true
+			break
+		}
+	}
+
+	if !found {
+		return fmt.Errorf("nested transaction not found")
+	}
+
+	for s.currentState() != id.state {
+		err := s.mergeIntoParent(s.EnforceInteractionLimits())
+		if err != nil {
+			return fmt.Errorf("cannot restart nested transaction: %w", err)
+		}
+	}
+
```

Comment:
I don't think it's necessary to do a full search just to get a better error message of it not existing instead of it not being at the top. Just checking that its not the top one would be ok I think.

Reply:
This is more about not accidentally merging nested transaction than about error messaging.

Suppose the calling provide an invalid nest transaction id.  Not checking the id first would cause us to accidently merge everything into the main transaction.

Reply:
I'll add a comment to clarify the intention here